src/v/cluster/partition.cc

```diff
@@ -113,7 +113,7 @@ ss::future<std::error_code> partition::prefix_truncate(
   model::offset rp_start_offset,
   kafka::offset kafka_start_offset,
   ss::lowres_clock::time_point deadline) {
-    if (!_log_eviction_stm) {
+    if (!_log_eviction_stm || !_raft->log_config().is_collectable()) {
```

Comment:
if the `_log_eviction_stm` is always going to be allocated at startup then we can remove the check from conditional

Reply:
There are few cases when it is still not allocated. I.e. internal topics.